src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1009,6 +1009,25 @@ public void release() {
 
     protected abstract void closeNoLock(String reason) throws ElasticsearchException;
 
+    public void flushAndClose() throws IOException {
+        if (isClosed.get() == false) {
+            logger.trace("flushAndClose now acquire writeLock");
+            try (ReleasableLock _ = writeLock.acquire()) {
+                logger.trace("flushAndClose now acquired writeLock");
+                try {
+                    logger.debug("flushing shard on close - this might take some time to sync files to disk");
+                    try {
+                        flush(); // TODO we might force a flush in the future since we have the write lock already even though recoveries are running.
+                    } catch (FlushNotAllowedEngineException ex) {
```

Comment:
I think it's important to bubble this up here we don't guarantee that this wont' throw this.